codalab/rest/groups.py

```diff
@@ -0,0 +1,162 @@
+
+
+@get('/groups/<group_spec>', apply=[AuthenticatedPlugin(), JsonApiPlugin()])
+def fetch_group(group_spec):
+    """Fetch a single group."""
+    group = get_group_info(group_spec, need_admin=False)
+    load_group_members(group)
+    document = GroupSchema().dump(group).data
+    include_group_relationships(document, [group])
+    return document
+
+
+@get('/groups', apply=[AuthenticatedPlugin(), JsonApiPlugin()])
+def fetch_groups():
+    """Fetch list of groups readable by the authenticated user."""
+    if request.user.user_id == local.model.root_user_id:
+        groups = local.model.batch_get_all_groups(
+            None,
+            {'user_defined': True},
+            None
+        )
+    else:
+        groups = local.model.batch_get_all_groups(
+            None,
+            {'owner_id': request.user.user_id, 'user_defined': True},
+            {'user_id': request.user.user_id}
+        )
+
+    for group in groups:
+        load_group_members(group)
+
+    document = GroupSchema(many=True).dump(groups).data
+    include_group_relationships(document, groups)
+    return document
+
+
+def load_group_members(group):
+    memberships = local.model.batch_get_user_in_group(group_uuid=group['uuid'])
+    group['admins'] = []
+    group['members'] = []
+    for m in memberships:
+        if m['user_id'] == group['owner_id']:
+            continue
+        elif m['is_admin']:
+            group['admins'].append(m['user_id'])
+        else:
+            group['members'].append(m['user_id'])
+
+
+def include_group_relationships(document, groups):
+    user_ids = set()
+    for group in groups:
+        user_ids.add(group['owner_id'])
+        user_ids.update(group['members'])
+        user_ids.update(group['admins'])
+    json_api_include(document, UserSchema(), local.model.get_users(user_ids))
+
+
+@delete('/groups', apply=[AuthenticatedPlugin(), JsonApiPlugin()])
+def delete_groups():
+    """Delete groups."""
+    group_ids = get_resource_ids(request.json, 'groups')
+
+    # Check permissions first
+    for group_id in group_ids:
+        get_group_info(group_id, need_admin=True)
+
+    # Delete groups
+    for group_id in group_ids:
+        local.model.delete_group(group_id)
+
+    abort(httplib.NO_CONTENT)
+
+
+# TODO(sckoo): support bulk create
+@post('/groups', apply=[AuthenticatedPlugin(), JsonApiPlugin()])
+def create_group():
+    """Create a group."""
+    group = GroupSchema(strict=True).load(request.json, partial=True).data
+    ensure_unused_group_name(group['name'])
+    group['owner_id'] = request.user.user_id
+    group = local.model.create_group(group)
+    local.model.add_user_in_group(request.user.user_id, group['uuid'], True)
+    return GroupSchema().dump(group).data
+
+
+@post('/groups/<group_spec>/relationships/admins',
+      apply=[AuthenticatedPlugin(), JsonApiPlugin()])
+def add_group_admins(group_spec):
+    user_ids = get_resource_ids(request.json, 'users')
```

Comment:
These 5 lines are copied between add_group_admins and add_group_members. Please don't have so much duplicate code.